<a href="https://colab.research.google.com/github/fullbread9977/Seoul-Bike-data-analysis/blob/main/%EB%94%B0%EB%A6%89%EC%9D%B4_%EC%9D%BC%EB%B3%84_%EB%B6%84%EC%84%9D_1~5%EC%9B%94_ipynb%EC%9D%98_%EC%82%AC%EB%B3%B8.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


구글 드라이브 런타임

폰트사용을 위해 설치하는 코드 


In [ ]:
import pandas as pd
import pandas_profiling
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from matplotlib import font_manager, rc
plt.rcParams['font.family'] = 'NanumGothic'

**필요한 라이브러리 설치**   
- pandas : 엑셀같은 표를 그리기위해 사용되는 패키지
- pandas_profiling : 데이터의 상태를 보기위해 사용되는 패키지
- numpy : 행렬 연산에 편의성을 제공하는 패키지 
- matplotlib.pyplot : 데이터 시각화, 여러가지 표를 그리기 위해 사용하는 패키지
- seaborn : 데이터분포 시각화를 위한 패키지
- matplotlib: 그래프의 rc에 한글 사용을 위한 패키지
- '남산고딕'이라는 한글 폰트를 사용


In [ ]:
df = pd.read_csv("/content/drive/MyDrive/seoulbiohub/공공자전거 이용정보(일별)_2020.01_05.csv", encoding="cp949")
df

대여일자  대여소번호                  대여소  ...   탄소량 이동거리(M) 이용시간(분)
0        2020-01-01    101     101. (구)합정동 주민센터  ...  0.00     0.0      29
1        2020-01-01    101     101. (구)합정동 주민센터  ...  0.75  3240.0      52
2        2020-01-01    101     101. (구)합정동 주민센터  ...  0.75  3250.0      21
3        2020-01-01    101     101. (구)합정동 주민센터  ...  1.59  6860.0      46
4        2020-01-01    101     101. (구)합정동 주민센터  ...  0.23   990.0       5
...             ...    ...                  ...  ...   ...     ...     ...
3052548  2020-05-31   3560  3560.성동구 견인차량 보관소 앞  ...  0.00     0.0      15
3052549  2020-05-31   3560  3560.성동구 견인차량 보관소 앞  ...  0.00     0.0      34
3052550  2020-05-31   3560  3560.성동구 견인차량 보관소 앞  ...  0.00     0.0      13
3052551  2020-05-31   3560  3560.성동구 견인차량 보관소 앞  ...  0.00     0.0      20
3052552  2020-05-31   3560  3560.성동구 견인차량 보관소 앞  ...  0.00     0.0      14

[3052553 rows x 11 columns]

공공데이터 5월 사용, 데이터 3,052,553 건 , 11 개의 컬럼

pr = df.profile_report()

pr

In [ ]:
import matplotlib.pyplot as plt
import seaborn as sns

데이터들의 상관관계를 보기위해 heatmap 라이브러리 로드

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3052553 entries, 0 to 3052552
Data columns (total 11 columns):
 #   Column   Dtype  
---  ------   -----  
 0   대여일자     object 
 1   대여소번호    int64  
 2   대여소      object 
 3   대여구분코드   object 
 4   성별       object 
 5   연령대코드    object 
 6   이용건수     int64  
 7   운동량      object 
 8   탄소량      object 
 9   이동거리(M)  float64
 10  이용시간(분)  int64  
dtypes: float64(1), int64(3), object(7)
memory usage: 256.2+ MB


In [ ]:
def clean(x):
    list(x)
    return x.strip().replace("\n","").replace('\\N',"")

In [ ]:
df

대여일자  대여소번호                  대여소  ...   탄소량 이동거리(M) 이용시간(분)
0        2020-01-01    101     101. (구)합정동 주민센터  ...  0.00     0.0      29
1        2020-01-01    101     101. (구)합정동 주민센터  ...  0.75  3240.0      52
2        2020-01-01    101     101. (구)합정동 주민센터  ...  0.75  3250.0      21
3        2020-01-01    101     101. (구)합정동 주민센터  ...  1.59  6860.0      46
4        2020-01-01    101     101. (구)합정동 주민센터  ...  0.23   990.0       5
...             ...    ...                  ...  ...   ...     ...     ...
3052548  2020-05-31   3560  3560.성동구 견인차량 보관소 앞  ...  0.00     0.0      15
3052549  2020-05-31   3560  3560.성동구 견인차량 보관소 앞  ...  0.00     0.0      34
3052550  2020-05-31   3560  3560.성동구 견인차량 보관소 앞  ...  0.00     0.0      13
3052551  2020-05-31   3560  3560.성동구 견인차량 보관소 앞  ...  0.00     0.0      20
3052552  2020-05-31   3560  3560.성동구 견인차량 보관소 앞  ...  0.00     0.0      14

[3052553 rows x 11 columns]

In [ ]:
df['운동량']

0            0.00
1           80.54
2           72.07
3          157.56
4           39.99
            ...  
3052548      0.00
3052549      0.00
3052550      0.00
3052551      0.00
3052552      0.00
Name: 운동량, Length: 3052553, dtype: object

In [ ]:
df['운동량'] = df['운동량'].apply(lambda x : clean(x) )

In [ ]:
df['운동량'] = pd.to_numeric(df['운동량']) 

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3052553 entries, 0 to 3052552
Data columns (total 11 columns):
 #   Column   Dtype  
---  ------   -----  
 0   대여일자     object 
 1   대여소번호    int64  
 2   대여소      object 
 3   대여구분코드   object 
 4   성별       object 
 5   연령대코드    object 
 6   이용건수     int64  
 7   운동량      float64
 8   탄소량      object 
 9   이동거리(M)  float64
 10  이용시간(분)  int64  
dtypes: float64(2), int64(3), object(6)
memory usage: 256.2+ MB


In [ ]:
df['탄소량'] = df['탄소량'].apply(lambda x : clean(x) )

In [ ]:
df['탄소량'] = pd.to_numeric(df['탄소량']) 

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3052553 entries, 0 to 3052552
Data columns (total 11 columns):
 #   Column   Dtype  
---  ------   -----  
 0   대여일자     object 
 1   대여소번호    int64  
 2   대여소      object 
 3   대여구분코드   object 
 4   성별       object 
 5   연령대코드    object 
 6   이용건수     int64  
 7   운동량      float64
 8   탄소량      float64
 9   이동거리(M)  float64
 10  이용시간(분)  int64  
dtypes: float64(3), int64(3), object(5)
memory usage: 256.2+ MB


In [ ]:
df.isna().sum()

대여일자            0
대여소번호           0
대여소             0
대여구분코드          0
성별         325967
연령대코드           0
이용건수            0
운동량          9166
탄소량          9166
이동거리(M)         0
이용시간(분)         0
dtype: int64

In [ ]:
df.dropna()
df = df.dropna()

In [ ]:
df.isna().sum()

대여일자       0
대여소번호      0
대여소        0
대여구분코드     0
성별         0
연령대코드      0
이용건수       0
운동량        0
탄소량        0
이동거리(M)    0
이용시간(분)    0
dtype: int64

In [ ]:
df.head(30)

대여일자  대여소번호               대여소  ...    탄소량  이동거리(M) 이용시간(분)
0   2020-01-01    101  101. (구)합정동 주민센터  ...   0.00      0.0      29
1   2020-01-01    101  101. (구)합정동 주민센터  ...   0.75   3240.0      52
2   2020-01-01    101  101. (구)합정동 주민센터  ...   0.75   3250.0      21
3   2020-01-01    101  101. (구)합정동 주민센터  ...   1.59   6860.0      46
4   2020-01-01    101  101. (구)합정동 주민센터  ...   0.23    990.0       5
5   2020-01-01    101  101. (구)합정동 주민센터  ...   0.23    990.0       5
6   2020-01-01    101  101. (구)합정동 주민센터  ...   0.14    610.0       2
7   2020-01-01    102   102. 망원역 1번출구 앞  ...   0.35   1520.0      65
8   2020-01-01    102   102. 망원역 1번출구 앞  ...   1.91   8230.0      40
9   2020-01-01    102   102. 망원역 1번출구 앞  ...   1.38   5940.0      52
10  2020-01-01    102   102. 망원역 1번출구 앞  ...   2.22   9580.0     157
11  2020-01-01    102   102. 망원역 1번출구 앞  ...   0.95   4080.0      75
12  2020-01-01    102   102. 망원역 1번출구 앞  ...   0.00      0.0      10
13  2020-01-01    102   102. 망원역 1번출구 앞  ...   0.31   1330.0       7
15  2020-01-01    102   102. 망원역 1번출구 앞  ...   0.12    500.0       5
16  2020-01-01    102   102. 망원역 1번출구 앞  ...   1.18   5050.0      38
17  2020-01-01    102   102. 망원역 1번출구 앞  ...   0.64   2750.0      11
18  2020-01-01    102   102. 망원역 1번출구 앞  ...   0.63   2730.0      66
19  2020-01-01    102   102. 망원역 1번출구 앞  ...  15.03  64810.0      66
20  2020-01-01    102   102. 망원역 1번출구 앞  ...   0.26   1110.0       9
21  2020-01-01    103   103. 망원역 2번출구 앞  ...   0.22    950.0       5
22  2020-01-01    103   103. 망원역 2번출구 앞  ...   0.54   2310.0      22
23  2020-01-01    103   103. 망원역 2번출구 앞  ...   0.20    870.0       6
24  2020-01-01    103   103. 망원역 2번출구 앞  ...   1.95   8420.0      48
25  2020-01-01    103   103. 망원역 2번출구 앞  ...   1.29   5580.0      67
26  2020-01-01    103   103. 망원역 2번출구 앞  ...   0.22    960.0       3
28  2020-01-01    103   103. 망원역 2번출구 앞  ...   2.06   8900.0      78
29  2020-01-01    103   103. 망원역 2번출구 앞  ...   1.17   5020.0     118
30  2020-01-01    103   103. 망원역 2번출구 앞  ...   0.67   2920.0      18
31  2020-01-01    103   103. 망원역 2번출구 앞  ...   0.43   1840.0      17

[30 rows x 11 columns]

In [ ]:
df['성별'] = df['성별'].apply(lambda x : clean(x) )

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [ ]:
df2 = df[(df['성별'] == 'M') | (df['성별'] == 'F')]   # 성별이 M 혹은 F인 것에 해당하는 행만 추출
df2

대여일자  대여소번호                  대여소  ...   탄소량 이동거리(M) 이용시간(분)
2        2020-01-01    101     101. (구)합정동 주민센터  ...  0.75  3250.0      21
3        2020-01-01    101     101. (구)합정동 주민센터  ...  1.59  6860.0      46
4        2020-01-01    101     101. (구)합정동 주민센터  ...  0.23   990.0       5
5        2020-01-01    101     101. (구)합정동 주민센터  ...  0.23   990.0       5
6        2020-01-01    101     101. (구)합정동 주민센터  ...  0.14   610.0       2
...             ...    ...                  ...  ...   ...     ...     ...
3052543  2020-05-31   3560  3560.성동구 견인차량 보관소 앞  ...  0.00     0.0     113
3052549  2020-05-31   3560  3560.성동구 견인차량 보관소 앞  ...  0.00     0.0      34
3052550  2020-05-31   3560  3560.성동구 견인차량 보관소 앞  ...  0.00     0.0      13
3052551  2020-05-31   3560  3560.성동구 견인차량 보관소 앞  ...  0.00     0.0      20
3052552  2020-05-31   3560  3560.성동구 견인차량 보관소 앞  ...  0.00     0.0      14

[1455636 rows x 11 columns]

In [ ]:
df2

대여일자  대여소번호                  대여소  ...   탄소량 이동거리(M) 이용시간(분)
2        2020-01-01    101     101. (구)합정동 주민센터  ...  0.75  3250.0      21
3        2020-01-01    101     101. (구)합정동 주민센터  ...  1.59  6860.0      46
4        2020-01-01    101     101. (구)합정동 주민센터  ...  0.23   990.0       5
5        2020-01-01    101     101. (구)합정동 주민센터  ...  0.23   990.0       5
6        2020-01-01    101     101. (구)합정동 주민센터  ...  0.14   610.0       2
...             ...    ...                  ...  ...   ...     ...     ...
3052543  2020-05-31   3560  3560.성동구 견인차량 보관소 앞  ...  0.00     0.0     113
3052549  2020-05-31   3560  3560.성동구 견인차량 보관소 앞  ...  0.00     0.0      34
3052550  2020-05-31   3560  3560.성동구 견인차량 보관소 앞  ...  0.00     0.0      13
3052551  2020-05-31   3560  3560.성동구 견인차량 보관소 앞  ...  0.00     0.0      20
3052552  2020-05-31   3560  3560.성동구 견인차량 보관소 앞  ...  0.00     0.0      14

[1455636 rows x 11 columns]

In [ ]:
df2 = pd.get_dummies(df2, columns= ['연령대코드', '성별'])

In [ ]:
df2

대여일자  대여소번호                  대여소  ... 연령대코드_AGE_008  성별_F  성별_M
2        2020-01-01    101     101. (구)합정동 주민센터  ...             0     1     0
3        2020-01-01    101     101. (구)합정동 주민센터  ...             0     1     0
4        2020-01-01    101     101. (구)합정동 주민센터  ...             0     0     1
5        2020-01-01    101     101. (구)합정동 주민센터  ...             0     0     1
6        2020-01-01    101     101. (구)합정동 주민센터  ...             1     0     1
...             ...    ...                  ...  ...           ...   ...   ...
3052543  2020-05-31   3560  3560.성동구 견인차량 보관소 앞  ...             0     0     1
3052549  2020-05-31   3560  3560.성동구 견인차량 보관소 앞  ...             0     0     1
3052550  2020-05-31   3560  3560.성동구 견인차량 보관소 앞  ...             0     0     1
3052551  2020-05-31   3560  3560.성동구 견인차량 보관소 앞  ...             0     0     1
3052552  2020-05-31   3560  3560.성동구 견인차량 보관소 앞  ...             0     0     1

[1455636 rows x 19 columns]

In [ ]:
val_name = list(df2['대여구분코드'].unique())
val_name

['정기', '일일(회원)', '단체', '일일(비회원)', 'BIL_021']

In [ ]:
df2 = df2.replace(val_name, ['memday','seasonal','cash','group','day'])

In [ ]:
colname = list(df2.columns)
colname

['대여일자',
 '대여소번호',
 '대여소',
 '대여구분코드',
 '이용건수',
 '운동량',
 '탄소량',
 '이동거리(M)',
 '이용시간(분)',
 '연령대코드_AGE_001',
 '연령대코드_AGE_002',
 '연령대코드_AGE_003',
 '연령대코드_AGE_004',
 '연령대코드_AGE_005',
 '연령대코드_AGE_006',
 '연령대코드_AGE_007',
 '연령대코드_AGE_008',
 '성별_F',
 '성별_M']

In [ ]:
df3 = df2[['대여구분코드',
 '이용건수',
 '운동량',
 '탄소량',
 '이동거리(M)',
 '이용시간(분)',
 '연령대코드_AGE_001',
 '연령대코드_AGE_002',
 '연령대코드_AGE_003',
 '연령대코드_AGE_004',
 '연령대코드_AGE_005',
 '연령대코드_AGE_006',
 '연령대코드_AGE_007',
 '연령대코드_AGE_008',
 '성별_F',
 '성별_M']]

In [ ]:
df3.head()

대여구분코드  이용건수     운동량   탄소량  ...  연령대코드_AGE_007  연령대코드_AGE_008  성별_F  성별_M
2  memday     1   72.07  0.75  ...              0              0     1     0
3  memday     1  157.56  1.59  ...              0              0     1     0
4  memday     1   39.99  0.23  ...              0              0     0     1
5  memday     1   25.48  0.23  ...              0              0     0     1
6  memday     1   17.63  0.14  ...              0              1     0     1

[5 rows x 16 columns]

In [ ]:
df3.to_csv("bycycle_daily.csv", encoding="utf-8")

In [ ]:
from sklearn.model_selection import train_test_split

In [ ]:
X = df3.drop('대여구분코드', axis=1)
y = df3['대여구분코드']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=100)

# 로지스틱 회귀분석을 통한 정기권 구매 예측

In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, confusion_matrix

In [ ]:
model = LogisticRegression()

In [ ]:
pip install h5py==2.10.0 

**에러 메세지**    
오류: pip의 종속성 해석기는 현재 설치된 모든 패키지를 고려하지 않습니다. 이 동작은 다음 종속성 충돌의 원인입니다.
tensorflow 2.6.0에는 h5py~=3.1.0이 필요하지만 호환되지 않는 h5py 2.10.0이 있습니다.

In [ ]:
model.fit(X_train, y_train)

/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

AttributeError: 'str' object has no attribute 'decode 의 경우   
h5py가 3 버전일 때 발생하는 에러이므로   
바로 위에 model = LogisticRegression() 을 누르고 +코드를 눌러 코드 칸을 추가한다.   
pip install h5py==2.10.0 입력하여 다운그레이드 해준다.

In [ ]:
pred = model.predict(X_test)

In [ ]:
accuracy_score(y_test, pred)

0.7447995383473935

In [ ]:
confusion_matrix(y_test, pred)

array([[     0,      0,      0,   3339,      0],
       [     0,      0,      0,     18,      0],
       [     0,      0,      0,    572,      0],
       [     0,      0,      0, 216832,      0],
       [     0,      0,      0,  70367,      0]])

# 디시즌 트리를 통한 이용권 구매 예측

In [ ]:
from sklearn.tree import DecisionTreeClassifier

In [ ]:
model =  DecisionTreeClassifier(max_depth =10)

In [ ]:
model.fit(X_train, y_train)

DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='gini',
                       max_depth=10, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=None, splitter='best')

In [ ]:
pred = model.predict(X_test)

In [ ]:
y_test

1018537      memday
2473591      memday
2187189    seasonal
1351449    seasonal
1584534      memday
             ...   
2084582    seasonal
2460769    seasonal
520236       memday
349052       memday
2613147      memday
Name: 대여구분코드, Length: 291128, dtype: object

In [ ]:
print(y_test, pred)

1018537      memday
2473591      memday
2187189    seasonal
1351449    seasonal
1584534      memday
             ...   
2084582    seasonal
2460769    seasonal
520236       memday
349052       memday
2613147      memday
Name: 대여구분코드, Length: 291128, dtype: object ['memday' 'memday' 'seasonal' ... 'memday' 'memday' 'memday']


In [ ]:
accuracy_score(y_test, pred)

0.7842632793822648

# 트리의 깊이를 조절하며 최적값 탐색 depth = 12일 때 값이 가장 높음

In [ ]:
for i in range(2, 20):
    model = DecisionTreeClassifier(max_depth = i)
    model.fit(X_train, y_train)
    pred = model.predict(X_test)
    print(i, round(accuracy_score(y_test, pred), 4))

2 0.7448
3 0.7575
4 0.763
5 0.7716
6 0.7783
7 0.7788
8 0.7819
9 0.7836
10 0.7842
11 0.7846
12 0.7848
13 0.7838
14 0.7821
15 0.7803
16 0.7772
17 0.7746
18 0.7707
19 0.7672


In [ ]:
model =  DecisionTreeClassifier(max_depth =12)  # depth= 12로 다시 분석

In [ ]:
model.fit(X_train, y_train)

DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='gini',
                       max_depth=12, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=None, splitter='best')

In [ ]:
pred = model.predict(X_test)

In [ ]:
accuracy_score(y_test, pred)

0.7847888214118875

정확도 78.48

In [ ]:
confusion_matrix(y_test, pred)

array([[    29,      0,      1,   2237,   1072],
       [     0,      0,      0,     17,      1],
       [     0,      0,    478,     82,     12],
       [    55,      0,     93, 200059,  16625],
       [    64,      0,     35,  42360,  27908]])

In [ ]:
from sklearn.tree import plot_tree

In [ ]:
!sudo apt-get install -y fonts-nanum
!sudo fc-cache -fv
!rm ~/.cache/matplotlib -rf

Reading package lists... Done
Building dependency tree       
Reading state information... Done
fonts-nanum is already the newest version (20170925-1).
0 upgraded, 0 newly installed, 0 to remove and 37 not upgraded.
/usr/share/fonts: caching, new cache contents: 0 fonts, 1 dirs
/usr/share/fonts/truetype: caching, new cache contents: 0 fonts, 3 dirs
/usr/share/fonts/truetype/humor-sans: caching, new cache contents: 1 fonts, 0 dirs
/usr/share/fonts/truetype/liberation: caching, new cache contents: 16 fonts, 0 dirs
/usr/share/fonts/truetype/nanum: caching, new cache contents: 10 fonts, 0 dirs
/usr/local/share/fonts: caching, new cache contents: 0 fonts, 0 dirs
/root/.local/share/fonts: skipping, no such directory
/root/.fonts: skipping, no such directory
/var/cache/fontconfig: cleaning cache directory
/root/.cache/fontconfig: not cleaning non-existent cache directory
/root/.fontconfig: not cleaning non-existent cache directory
fc-cache: succeeded


폰트 설치를 위한 코드

https://teddylee777.github.io/colab/colab-korean

In [ ]:
#plt.rc('font', family='NanumGothic.ttf') 
plt.figure(figsize=(20,10))
plot_tree(model, feature_names=X_train.columns, fontsize=15, label ="None", max_depth = 5)

[Text(558.0, 504.7714285714286, '연령대코드_AGE_002 <= 0.5\n0.386\n1164508\n[13168, 76, 2177, 868397, 280690]'),
 Text(279.0, 427.11428571428576, '이용건수 <= 1.5\n0.328\n776059\n[8841, 43, 2125, 619177, 145873]'),
 Text(139.5, 349.4571428571429, '이용시간(분) <= 28.5\n0.388\n454037\n[1639, 39, 2007, 336626, 113726]'),
 Text(69.75, 271.8, '연령대코드_AGE_003 <= 0.5\n0.291\n283037\n[464, 38, 597, 233392, 48546]'),
 Text(34.875, 194.14285714285717, '연령대코드_AGE_001 <= 0.5\n0.218\n192793\n[344, 19, 591, 168968, 22871]'),
 Text(17.4375, 116.48571428571432, '연령대코드_AGE_004 <= 0.5\n0.175\n166161\n[224, 14, 591, 150138, 15194]'),
 Text(8.71875, 38.82857142857142, '\n  (...)  \n'),
 Text(26.15625, 38.82857142857142, '\n  (...)  \n'),
 Text(52.3125, 116.48571428571432, '이용시간(분) <= 13.5\n0.417\n26632\n[120, 5, 0, 18830, 7677]'),
 Text(43.59375, 38.82857142857142, '\n  (...)  \n'),
 Text(61.03125, 38.82857142857142, '\n  (...)  \n'),
 Text(104.625, 194.14285714285717, '이용시간(분) <= 10.5\n0.409\n90244\n[120, 19, 6, 64424

RuntimeWarning: Glyph 47049 missing from current font.

런타임 경고: 현재 글꼴에서 Glyph 47049가 누락되었습니다.      
외부 API에서 가져와 폰트를 설치시도    
설치가 제대로 됬다고 컴파일 됬으나 인식하지 못하는 3번이상의 오류,    
구글 드라이브에 폰트를 넣어야되는 걸로 예상됨
- 폰트를 넣고나서 경로지정 후 다시해보기
- Mydrive 안에 폰트를 넣고 다시 해봤으나 오류..

# 랜덤포레스트

In [ ]:
from sklearn.ensemble import RandomForestClassifier

In [ ]:
model = RandomForestClassifier(max_depth = 10, random_state = 100)

In [ ]:
model.fit(X_train, y_train)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=10, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=100,
                       n_jobs=None, oob_score=False, random_state=100,
                       verbose=0, warm_start=False)

In [ ]:
pred = model.predict(X_test)

In [ ]:
y_test

1018537      memday
2473591      memday
2187189    seasonal
1351449    seasonal
1584534      memday
             ...   
2084582    seasonal
2460769    seasonal
520236       memday
349052       memday
2613147      memday
Name: 대여구분코드, Length: 291128, dtype: object

In [ ]:
accuracy_score(y_test, pred)

0.7780082987551867

In [ ]:
confusion_matrix(y_test, pred)

array([[     0,      0,      1,   2603,    735],
       [     0,      0,      0,     18,      0],
       [     0,      0,    393,    169,     10],
       [     0,      0,     22, 207241,   9569],
       [     0,      0,     29,  51472,  18866]])

# 최적 max_depth 찾기

In [ ]:
for i in range(2,20):
    model = RandomForestClassifier(max_depth=i, random_state=100)
    model.fit(X_train, y_train)
    pred = model.predict(X_test)
    print(i, accuracy_score(y_test, pred))

2 0.7447995383473935
3 0.7447995383473935
4 0.7447995383473935
5 0.7473928993432442
6 0.7562275013052678
7 0.7629221510813113
8 0.7661784507158363
9 0.7712140364375807
10 0.7780082987551867
11 0.7811752905938282
12 0.7833530268472975
13 0.7846617295485148
14 0.785640680388008
15 0.7865406281772965
16 0.7868909895303784
17 0.7869596878349042
18 0.7866402407188591
19 0.7861524827567256


17에서 78.69%의 정확도로 가장높게 나오고 18번 째 부터 감소하여 종료
- 오류로 종료가 아님

시간이 너무 오래걸려서 12까지만 돌리고 종료...

- 가지가 늘어날수록 제곱으로 늘어나기 때문에 느려질 수 밖에 없다



In [ ]:
model = RandomForestClassifier(max_depth=17, random_state=100)
model.fit(X_train, y_train)
pred = model.predict(X_test)
print( accuracy_score(y_test, pred))

0.7869596878349042


17번째로 값을 수정하고 코드실행

In [ ]:
from sklearn.metrics import log_loss

In [ ]:
#logloss = log_loss(y_test, pred)

In [ ]:
confusion_matrix(y_test, pred)

array([[    23,      0,      1,   2253,   1062],
       [     0,      0,      0,     14,      4],
       [     0,      0,    478,     78,     16],
       [    17,      0,     61, 201333,  15421],
       [    21,      0,     37,  43037,  27272]])

In [ ]:
accuracy_score(y_test, pred)

0.7869596878349042

정확도 78.69%

In [ ]:
from sklearn.metrics import classification_report

In [ ]:
print(classification_report(y_test, pred))

/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1272: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


              precision    recall  f1-score   support

        cash       0.38      0.01      0.01      3339
         day       0.00      0.00      0.00        18
       group       0.83      0.84      0.83       572
      memday       0.82      0.93      0.87    216832
    seasonal       0.62      0.39      0.48     70367

    accuracy                           0.79    291128
   macro avg       0.53      0.43      0.44    291128
weighted avg       0.76      0.79      0.76    291128

